SCons/Tool/msvs.py

```diff
@@ -1465,22 +1480,35 @@ def Build(self):
 
         _GenerateV10User.Build(self)
 
+def _projectDSPNodes(env):
+    if 'projects' not in env:
+        raise SCons.Errors.UserError("You must specify a 'projects' argument to create an MSVSSolution.")
+    projects = env['projects']
+    if not SCons.Util.is_List(projects):
+        raise SCons.Errors.InternalError("The 'projects' argument must be a list of nodes.")
+    projects = SCons.Util.flatten(projects)
+    if len(projects) < 1:
+        raise SCons.Errors.UserError("You must specify at least one project to create an MSVSSolution.")
+    sln_suffix = os.path.normcase(env.subst('$MSVSSOLUTIONSUFFIX'))
+    dspnodes = []
+    for p in projects:
+        node = env.File(p)
+        if os.path.normcase(str(node)).endswith(sln_suffix):
```

Comment:
Should this yield an error?

Reply:
It think it is easier on an end-user if it does not yield an error and simply ignores the solution files.

The following example comes from the template in TestSConsMSVS.py (i.e., ignore the template string replacement).

Likely usage of multiple projects and an explicit solution by a user is:
```
p1 = env.MSVSProject(
    target = '%(PROJECT_FILE_1)s',
    MSVS_PROJECT_GUID = '%(PROJECT_GUID_1)s',
    slnguid = '{SLNGUID}',
    srcs = testsrc,
    incs = testincs,
    localincs = testlocalincs,
    resources = testresources,
    misc = testmisc,
    buildtarget = 'Test_1.exe',
    variant = 'Release',
    auto_build_solution = %(AUTOBUILD_SOLUTION)s,
)

p2 = env.MSVSProject(
    target = '%(PROJECT_FILE_2)s',
    MSVS_PROJECT_GUID = '%(PROJECT_GUID_2)s',
    slnguid = '{SLNGUID}',
    srcs = testsrc,
    incs = testincs,
    localincs = testlocalincs,
    resources = testresources,
    misc = testmisc,
    buildtarget = 'Test_2.exe',
    variant = 'Release',
    auto_build_solution = %(AUTOBUILD_SOLUTION)s,
)

env.MSVSSolution(
    target = '%(SOLUTION_FILE)s',
    projects = [p1, p2],
    variant = 'Release',
)
```

Outcomes:
* When `auto_build_solution = 1`, MSVSProject is returning multiple nodes: the project node and the solution node.

* When `auto_build_solution = 0`, MSVSProject is returning the project node.

Passing the return values from MSVSProject directly to MSVSSolution seems easier than making the user explicitly filter out the solution nodes from the projects list.

I only tested opening a solution with both project and solution file references in the Projects section with VS2022.  VS2022 complained upon opening but I think it still allowed the solution/project files to be used.  I don't remember exact sequence.

Reply:
Note: the default value for `auto_build_solution` is `1`.

In the example above, if the `auto_build_solution` keyword argument were not present, the solution file node would be returned with the project node.

Depending on the experience level of the user, raising an error might be confusing.

Reply:
I don't think any other builders ignore an incorrect source file..  

I'd suggest error out and give a useful hint.  That's be the most consistent, and ignoring things silently isn't really the SCons way..

Somethink like, you've included a solution file as as source to MSVSSolution(), Likely you want to set `auto_build_solution=0` in your call to MSVSProject()?

In the case where `auto_build_solution` is set (as above), is the per project solution file ever reasonably going to be used?

Reply:
This has been lurking in the code since version 1.0.  How could I possibly be the first one to try and use it?

> In the case where `auto_build_solution` is set (as above), is the per project solution file ever reasonably going to be used?

In general, probably not.

In far-fetched examples perhaps it depends on the configuration of the individual projects. For example:
* A debug configuration and a release configuration.  Other than a code browser, the release configuration isn't doing much.  The solution with only the debug configuration would be a code browser and debugging solution (as would the "main" solution file).
* A configuration targeting x86_64 and a configuration targeting arm64.  On an x86_64 box, the arm64 configuration is really only useful as a code browser.  The x86_64 configuration solution file (and the "main" solution file) would be useful as a code browser and debugging solution.

I may not be able to get to this very quickly.

Reply:
This describes the behavior with solution files in the project file sections:
https://github.com/SCons/scons/issues/4613#issuecomment-2410801845

Feedback when the issue were described, the proposed solution in this PR was mentioned, and an unanswered question was asked would have been helpful.